File: examples/CBuildr3/ExTblOdU.cpp
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include <stdio.h>

#include "ExTblOdU.h"
//---------------------------------------------------------------------------
#pragma link "OvcTable"
#pragma link "OvcTCmmn"
#pragma link "OvcBase"
#pragma link "OvcTCEdt"
#pragma link "OvcTCStr"
#pragma link "OvcTCell"
#pragma link "OvcTCHdr"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
  S = "Test";
}
//---------------------------------------------------------------------------
void __fastcall TForm1::TBLGetCellData(TObject *Sender, int RowNum, int ColNum,
  Pointer &Data, TOvcCellDataPurpose Purpose)
{
  Data = &S;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcTCString1OwnerDraw(TObject *Sender,
  TCanvas *TableCanvas, const TRect &CellRect, int RowNum, int ColNum,
  const TOvcCellAttributes &CellAttr, Pointer Data, bool &DoneIt)
{
  bool F;
  TRect R;
  TCanvas* C;
  char Buf[255];

  DoneIt = True;
  R = CellRect;
  C = TableCanvas;

  // paint background
  C->Brush->Style = bsClear;
  C->Pen->Color = clBtnFace;
  C->Rectangle(R.Left, R.Top, R.Right, R.Bottom);

  // draw button
  F = (RowNum == TBL->ActiveRow) && (ColNum == TBL->ActiveCol);
  DrawButtonFace(C, R, 1, bsNew, true, F, F);

  // draw data
  if ((RowNum % 2) == 1)
    C->Font->Color = clRed;
  else
    C->Font->Color = clBlack;
  sprintf(Buf, "%d:%d", RowNum, ColNum);
  DrawText(C->Handle, Buf, StrLen(Buf),
    (RECT*)&R, DT_CENTER | DT_VCENTER | DT_SINGLELINE);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::BitBtn1Click(TObject *Sender)
{
  Close();    
}
//---------------------------------------------------------------------------

File: examples/Delphi/EXTBLODU.pas
{$J+} {Writable constants}

unit ExTblODu;
  {-example showing owner-draw cells}

interface

uses
  WinTypes, WinProcs, StdCtrls, Buttons, Classes,
  Controls, Forms, Graphics, SysUtils,
  OvcBase, OvcTCmmn, OvcTable, OvcTCHdr,
  OvcTCell, OvcTCStr, OvcTCEdt;

type
  TForm1 = class(TForm)
    TBL: TOvcTable;
    OvcController1: TOvcController;
    OvcTCString1: TOvcTCString;
    OvcTCRowHead1: TOvcTCRowHead;
    OvcTCColHead1: TOvcTCColHead;
    BitBtn1: TBitBtn;
    procedure TBLGetCellData1(Sender: TObject; RowNum: Longint;
      ColNum: Integer; var Data: Pointer; Purpose: TOvcCellDataPurpose);
    procedure OvcTCString1OwnerDraw(Sender: TObject; TableCanvas: TCanvas;
      const CellRect: TRect; RowNum: Longint; ColNum: Integer;
      const CellAttr: TOvcCellAttributes; Data: Pointer;
      var DoneIt: Boolean);
    procedure BitBtn1Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}
const
  S : string[10] = 'Test';

procedure TForm1.TBLGetCellData1(Sender: TObject; RowNum: Longint;
  ColNum: Integer; var Data: Pointer; Purpose: TOvcCellDataPurpose);
begin
  Data := @S;
end;

procedure TForm1.OvcTCString1OwnerDraw(Sender: TObject;
  TableCanvas: TCanvas; const CellRect: TRect; RowNum: Longint;
  ColNum: Integer; const CellAttr: TOvcCellAttributes; Data: Pointer;
  var DoneIt: Boolean);
var
  F   : Boolean;
  R   : TRect;
  C   : TCanvas;
  Buf : array[0..255] of Char;
begin
  DoneIt := True;
  
  R := CellRect;
  C := TableCanvas;

  {paint background}
  C.Brush.Style := bsClear;
  C.Pen.Color := clBtnFace;
  C.Rectangle(R.Left, R.Top, R.Right, R.Bottom);

  {draw button}
  F := (RowNum=TBL.ActiveRow) and (ColNum=TBL.ActiveCol);
  DrawButtonFace(C, R, 1, bsNew, True, F, F);

  {draw data}
  if Odd(RowNum) then
    C.Font.Color := clRed
  else
    C.Font.Color := clBlack;
  StrPLCopy(Buf, Format('%d:%d', [RowNum, ColNum]), 255);
  DrawText(C.Handle, Buf, StrLen(Buf), R, DT_CENTER or DT_VCENTER or DT_SINGLELINE);
end;

procedure TForm1.BitBtn1Click(Sender: TObject);
begin
  Close;
end;

end.
